archive/app/controllers/ArchiveController.scala

```diff
@@ -66,9 +68,31 @@ object ArchiveController extends Controller with Logging with ExecutionContexts
     case _ => false
   }
 
-  private def destinationFor(path: String) = DynamoDB.destinationFor(path).map(_.filterNot { destination =>
-      linksToItself(path, destination.location)
-  })
+  def retainShortUrlCampaign(path: String)(destination: Destination): Destination = {
+    // if the path is a short url with a campaign, and the destination doesn't have a campaign, pass it through the redirect.
+    val shortUrlWithCampaign = """.*www\.theguardian\.com/p/[\w\d]+/([\w\d]+)$""".r
+    val urlWithCampaignParam = """.*www\.theguardian\.com.*?.*CMP=.*$""".r
+
+    val destinationHasCampaign = destination.location match {
+      case shortUrlWithCampaign(_) => true
+      case urlWithCampaignParam() => true
+      case _ => false
+    }
+
+    path match {
+      case shortUrlWithCampaign(campaign) if !destinationHasCampaign => {
+        val uri = javax.ws.rs.core.UriBuilder.fromPath(destination.location)
```

Comment:
Out of curiosity what made you choose this library? I just did a quick scan of the codebase (as I didn't know what library we used) and it seems `java.net.URI` is preferred?